### ISSUE TITLE
Remove hacky $ concatenation in HTML Print

### ISSUE THREAD

**Issue by danscime** [JacquesCarette/Drasil#1780]
Based on comments in: https://github.com/JacquesCarette/Drasil/commit/ac99c9d40835ca839973b25c5dc9c3685530075d

**Comment by JacquesCarette** [comment]
You should 'dispatch' to the routines of the other printer differently and directly.  In particular the `BlockEqn` shouldn't call the HTML `pExpr` but rather the latex one.

Also, based on other comments, I think we don't quite want to go 100% MathJax.  I think we want to print the simple stuff still in HTML, and only the complex stuff in latex.

Easiest way to do that would be to change the current HTML `pExpr` to return an `Either Doc Doc`. If it's a `Left` (say), we know it only contains plain HTML, so can be used as is; if it's a `Right`, it's still HTML, but with 'hacks', so that we'd rather go with MathJax in that case.

**Comment by danscime** [comment]
The biggest hurdle seems to be the type mismatches between the HTML and Tex Print files.

<img width="581" alt="Screen Shot 2019-07-31 at 11 35 22 AM" src="https://user-images.githubusercontent.com/28261339/62225945-5df7f980-b387-11e9-8449-5680327f5d7f.png">

<img width="566" alt="Screen Shot 2019-07-31 at 11 35 52 AM" src="https://user-images.githubusercontent.com/28261339/62225946-5df7f980-b387-11e9-8e08-961748e6481a.png">

I can not do a straightforward call to the Tex functions.


As for the `Either Doc Doc`, how do we want to decide what is a hack, and what is not?

**Comment by JacquesCarette** [comment]
Yep, you will need a `PrintingInformation` structure to be able to go between them. But that's how options are passed, so that's going to be needed long-term anyways.

The decision as to what is a hack or not is fully subjective. I would say that any kind of 2-dimensional printing (fractions, integrals) are a hack. We can review things further, and move items between the two categories easily if we have this infrastructure in place.

**Comment by danscime** [comment]
Related to #479

**Comment by danscime** [comment]
I'm still unsure of where to go from here. I can add the PrintingInformation parameters, but then how should be translating to or from `D` and `Doc`.

Currently I have a function that goes from `D` to `Doc`. But I don't know exactly how we want things to be done.

<img width="257" alt="Screen Shot 2019-07-31 at 4 19 13 PM" src="https://user-images.githubusercontent.com/28261339/62245181-08374780-b3af-11e9-828b-caef5b69ad54.png">

**Comment by danscime** [comment]
@samm82 and I were discussing. It seems like the Latex print uses a `D` type to stage the math for printing. HTML should also use this as it would need to stage the math.

So the HTML file may need to undergo some renovations in which the functions pass through PrintingInformation as well as outputting D types instead of Doc types.

Not sure how or where to start.

**Comment by Mornix** [comment]
Correct me if I'm wrong, but that seems like an incorrect approach. `D` is very much a LaTeX specific "thing" specifically designed to abstract away the mode switching. Wouldn't a more appropriate solution to be:

`printMath d = runPrint d Text`? 
If the printed text is just text, then no `$` should be generated, where as if it is truly math then the mode should be set (within `D` to `Math` and the `$` generated.

**Comment by danscime** [comment]
@Mornix That definitely may be the key to this issue! Will continue to work with this.

**Comment by JacquesCarette** [comment]
Note that, in this instance, we know more: when we encounter a `BlockEqn` while printing HTML, we **know** we're going to be in `Math` mode. Nevertheless, as @Mornix says, passing the whole `BlockEqn` directly to the latex printer (not just its contents) should in fact take care of adding the `$$`. The same analysis needs to happen in the 'other' case (`Expr` ?). But yeah, letting the latex printer figure it out, as it already has the mode logic built-in, is probably the thing to do.
